### DISCUSSION TITLE
DeFi protocol fuzzing for getters questions

### DISCUSSION THREAD

**Discussion by accurec** [Cyfrin/foundry-full-course-cu#3117] [Q&A] [answered]
Hello!

For the stablecoin DeFi protocol invariant fuzz testing Patrick mentioned that we need to test getters so that they never revert.

The code like this is understandable for me:
```
    function invariant_gettersShouldNotRevert() public view {
        dsce.getPrecision();
        dsce.getLiquidationBonus();
        dsce.getAdditionalFeedPrecision();
        dsce.getLiquidationPrecision();
    }
```

In the above example we're basically testing pure getters on DSCEngine. In the meantime, would it be possible to test something like `function getUsdValue(address token, uint256 amount) public view returns (uint256)` and input different parameters like fuzzing does? 

I tried to add parameter to the test, but then getting error:
```
FAIL: failed to set up invariant testing environment: Invariant test function should have no inputs
```

Let me know if something else can be done.

Thanks!

**Comment by EngrPips** [comment]
There are a number of tokens your Protocol support so fuzzing the `token` when trying to get the value of `USD` for any given amount does not actually make sense. You can fuzz the amount, and that makes total sense, but fuzzing the token does not make sense to me.

I don't understand what you mean when you said, `I tried to add parameter to the test, but then getting error:` Maybe show the code so I can be sure what you are saying.

**Reply by accurec** [reply]
I got the error when I changed test signature to something like this:
```
function invariant_gettersShouldNotRevert(uint256 tokenSeed) public view {
```

As for fuzzing tokens and amounts I think it all makes sense, because we want to test different scenarios and make sure that none of them make getters to revert.

**Reply by EngrPips** [reply, marked as answer]
You can't do this `function invariant_gettersShouldNotRevert(uint256 tokenSeed) public view {` as the `invariant` function does not take any arguments. You could create a handler and use the handler to give fuzz input to the getter function of the implementation contract.

For your argument that it makes to fuzz `token` input in a getter that should never revert, I leave you to do you.

**Reply by accurec** [reply]
That's what I thought, thank you!

**Reply by EngrPips** [reply]
You are welcome, Friend.
